### PR TITLE
feat: student test submission

### PR DESCRIPTION
**Pull request by joyce-shi**
## Notion ticket link

[Student Experience Submission](https://www.notion.so/uwblueprintexecs/StudentExperience-submission-94965d4fb55f4cf7b63b443fb19f5c5e?pvs=4)

## Implementation description
* handling for student assessment submission

## What should reviewers focus on?
* works as expected
* matches design (there's a couple discrepancies that are noted in the PR)

### CODE REVIEW

File: frontend/src/utils/StudentUtils.ts [hunk 1]
```diff
@@ -76,3 +76,13 @@ export const questionStatus = (
   if (answers[index].isCompleted) return QuestionNumberTypes.COMPLETED;
   return QuestionNumberTypes.UNATTEMPTED;
 };
+
+export const mapAnswersToResultsArray = (answers: Answers[]): number[][][] => {
+  return answers.map((answer) => {
+    const row: number[][] = [];
+    answer.elements.map((element) => {
```
Comment (jfdoming) [inline frontend/src/utils/StudentUtils.ts]: I think you should be able to do:

```suggestion
    answer.elements.forEach((element) => {
```
or replace the whole thing with:
```tsx
  return answers.map((answer) => answer.elements.map((element) => element.elementAnswers))
```

File: frontend/src/components/assessments/student-experience/NavButtons.tsx [hunk 2]
```diff
@@ -1,15 +1,53 @@
...
+    await submitTest({
+      variables: {
+        id: testSession?.id,
+        result: {
+          student: authenticatedUser?.id,
+          answers: mapAnswersToResultsArray(answers),
+        },
+      },
+    });
+    if (error) {
```
Comment (jfdoming) [inline frontend/src/components/assessments/student-experience/NavButtons.tsx]: Will this show an error as expected? Usually when you await something it will throw upon error, not sure about Apollo though...

edit yep https://stackoverflow.com/a/59472340

File: frontend/src/utils/StudentUtils.ts (hunk 1 above)
Comment (joyce-shi) [reply frontend/src/utils/StudentUtils.ts]: nice, thank you!

File: frontend/src/components/assessments/student-experience/question-elements/MultiSelect.tsx [hunk 3]
```diff
@@ -16,8 +18,8 @@ const MultiSelect = ({
 
   return (
     <CheckboxGroup
-      onChange={(e) => updateAnswer(e as number[])}
-      value={currentAnswer}
+      onChange={(e) => updateAnswer(stringOrNumberArrayToNumberArray(e))}
+      value={currentAnswer.map(String)}
```
Comment (joyce-shi) [inline frontend/src/components/assessments/student-experience/question-elements/MultiSelect.tsx:22]: i hate this. but from:
- https://github.com/chakra-ui/chakra-ui/issues/6966
- https://github.com/chakra-ui/chakra-ui/issues/4838

it looks like we have to cast from number -> string -> number for both `MultipleChoice` and `MultiSelect` components?

File: frontend/src/components/assessments/student-experience/SubmitButton.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    return answers.filter((answer) => !isCompleted(answer)).length;
+  }, [answers]);
+
+  const header = useMemo(() => {
+    const question = incompleteQuestionCount === 1 ? "Question" : "Questions";
+    if (incompleteQuestionCount) {
+      return `You have ${incompleteQuestionCount} Unanswered ${question}`;
+    }
+    return "Submit the test";
+  }, [incompleteQuestionCount]);
```
Comment (joyce-shi) [inline frontend/src/components/assessments/student-experience/SubmitButton.tsx:74]: a couple notes:
- in the designs the `${incompleteQuestionCount} Unanswered` text is red and bolded, any thoughts on the most elegant way to do this? it's kind of hard to adapt the existing modal component, but i'm hesitant to create a modal from scratch just for this as well
- in the designs, it shows the word version of the number, not the number itself (i.e `one` instead of `1`). is there a easy way to do this? i can't think of one... unless i'm overlooking something obvious

File: frontend/src/APIClients/mutations/TestSessionMutations.ts [hunk 5]
```diff
@@ -8,6 +8,14 @@ export const CREATE_TEST_SESSION = gql`
   }
 `;
 
+export const SUBMIT_TEST = gql`
+  mutation CreateTestSessionResult($id: ID!, $result: ResultRequestDTO!) {
```
Comment (jfdoming) [inline frontend/src/APIClients/mutations/TestSessionMutations.ts]: nit: it's not immediately clear what this is an ID for? maybe `testSessionId` or `studentId` or something?

File: frontend/src/components/assessments/student-experience/Instructions.tsx [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import React from "react";
+import { Text } from "@chakra-ui/react";
+
+const Instructions = (): React.ReactElement => {
+  return (
+    <Text color="grey.300" textStyle="subtitle1">
+      All responses will be autosaved. Make sure you answer all the questions
+      before submitting your test. You can also do the questions in ANY order
+      meaning that can skip a question and come back to it later!
```
Comment (jfdoming) [inline frontend/src/components/assessments/student-experience/Instructions.tsx]:
```suggestion
      meaning that you can skip a question and come back to it later!
```

File: frontend/src/components/assessments/student-experience/SubmitButton.tsx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        setIsLoading(false);
+        setIsSubmitted(true);
+      })
+      .catch(() => {
+        setIsLoading(false);
+        showToast({
+          message: "Assessment failed to submit. Please try again.",
+          status: "error",
+        });
+      });
```
Comment (jfdoming) [inline frontend/src/components/assessments/student-experience/SubmitButton.tsx]: nit: You can probably make use of async/await instead of then/catch
```suggestion
    try {
      await submitTest({
        variables: {
          id: testSession?.id,
          result: {
            student: authenticatedUser?.id,
            answers: mapAnswersToResultsArray(answers),
          },
        },
      });

      setIsSubmitted(true);
    } catch (e) {
      showToast({
        message: "Assessment failed to submit. Please try again.",
        status: "error",
      });
    }

    setIsLoading(false);
```

File: frontend/src/components/assessments/student-experience/SubmitButton.tsx (hunk 4 above)
Comment (jfdoming) [reply frontend/src/components/assessments/student-experience/SubmitButton.tsx:74]: For 1., you could probably just change the type of `header` (and `body`) to `React.ReactNode`, and pass in a React Element instead. i.e., you'd return something like `<>You have {incompleteQuestionCount} <Box color="red">Unanswered</Box>{question}</>;` (the other `return` could stay unchanged).

For 2., yeah idk, you could add a library for this if you wanted or just manually check, but I'd probably just use the # unless the designer cares.

File: frontend/src/components/pages/student/StudentRouting.tsx [hunk 8]
```diff
@@ -61,10 +56,6 @@ const StudentRouting = (): React.ReactElement => {
         setTestSession,
         className,
         setClassName,
-        currentQuestionIndex,
-        setCurrentQuestionIndex,
-        answers,
-        setAnswers,
```
Comment (jfdoming) [inline frontend/src/components/pages/student/StudentRouting.tsx:67]: 🗑️ love to see all the `-` symbols 🙌

File: frontend/src/components/assessments/TestSubmissionMessage.tsx [hunk 9]
```diff
@@ -1,17 +1,23 @@
...
+
   return (
     <Center minHeight="100vh">
       <MessageContainer
-        buttonRoute={HOME_PAGE}
         buttonText="Go to Home Page"
         image={TestSubmissionIllustration}
+        onClick={handleClick}
         paragraphs={[]}
         subtitle="Great Job! Your test submission was a success.
```
Comment (jfdoming) [inline frontend/src/components/assessments/TestSubmissionMessage.tsx]: Just noticed this, it's really a design thing, but given the demographic that will be reading this copy, I'm wondering if we could simplify the wording? Something like `Great job! We've saved your answers.` might be a little more understandable maybe?

File: frontend/src/components/assessments/TestSubmissionMessage.tsx [hunk 10]
```diff
@@ -1,17 +1,23 @@
...
 import TestSubmissionIllustration from "../../assets/illustrations/test-submission.svg";
-import { HOME_PAGE } from "../../constants/Routes";
+import AuthContext from "../../contexts/AuthContext";
 import MessageContainer from "../common/MessageContainer";
 
 const TestSubmissionMessage = (): React.ReactElement => {
+  const { setAuthenticatedUser } = useContext(AuthContext);
+
+  const handleClick = () => {
+    setAuthenticatedUser(null);
```
Comment (jfdoming) [inline frontend/src/components/assessments/TestSubmissionMessage.tsx:12]: Also wondering, I noticed that if you're logged in as a student and you refresh the page, it not only loses your test progress but also logs you out (sends you to the homepage). That behaviour seems a little annoying since they'd have to re-enter their name and code again. Any idea if there's a quick fix?

File: frontend/src/components/assessments/student-experience/question-elements/MultiSelect.tsx (hunk 3 above)
Comment (jfdoming) [reply frontend/src/components/assessments/student-experience/question-elements/MultiSelect.tsx:22]: Yeah I don't think there's a way around this unfortunately.... Probably the "standard" way to handle this is actually to store things only as strings on the backend (floating point is unreliable 💀) and only convert to `number` if we need to do some sort of numerical manipulation (or perhaps when checking answers, since we'd want to make sure `0.50` is marked as correct for `0.5`).

File: frontend/src/components/pages/student/WriteAssessmentPage.tsx [hunk 11]
```diff
@@ -1,43 +1,80 @@
...
+      }}
+    >
+      {isLoading && (
+        <LoadingState
+          fullPage
+          text="Please wait while we submit your assessment."
+        />
+      )}
+      {!isLoading && isSubmitted && <TestSubmissionMessage />}
+      {!isLoading && !isSubmitted && (
```
Comment (jfdoming) [inline frontend/src/components/pages/student/WriteAssessmentPage.tsx]: Personal preference would be 
```suggestion
      {isLoading ? (
        <LoadingState
          fullPage
          text="Please wait while we submit your assessment."
        />
      ) : (
        isSubmitted ? (
          ...
        ) : (
          ...
        )
      )}
```
but this way is fine too.

**Review by jfdoming — APPROVED** [review]
huge

File: frontend/src/components/assessments/student-experience/question-elements/MultiSelect.tsx (hunk 3 above)
Comment (joyce-shi) [reply frontend/src/components/assessments/student-experience/question-elements/MultiSelect.tsx:22]: LOL CS370 flashbacks 😂 re: backend update - i'm not sure if making that update is worth the effort right now, so do you think we are okay leaving it as is? or do you think we should create a ticket for the backlog?

File: frontend/src/components/assessments/TestSubmissionMessage.tsx (hunk 10 above)
Comment (joyce-shi) [reply frontend/src/components/assessments/TestSubmissionMessage.tsx:12]: yeah +1. i think carissa & i talked about this a bit before & we wanted to persist login + answers to local storage as a solution to this, so theres a ticket in the backlog for that.

for now, maybe i can add a prompt that shows on reload to warn them from leaving the page? i had it as a TODO for this PR, but forgot to do that

File: frontend/src/components/assessments/TestSubmissionMessage.tsx (hunk 9 above)
Comment (joyce-shi) [reply frontend/src/components/assessments/TestSubmissionMessage.tsx]: i like!

File: frontend/src/components/assessments/student-experience/question-elements/MultiSelect.tsx (hunk 3 above)
Comment (jfdoming) [reply frontend/src/components/assessments/student-experience/question-elements/MultiSelect.tsx:22]: yeah prob not worth the effort atm, only if it becomes a problem

File: frontend/src/components/assessments/TestSubmissionMessage.tsx (hunk 10 above)
Comment (jfdoming) [reply frontend/src/components/assessments/TestSubmissionMessage.tsx:12]: yeah whatever you think is best! it's not really related to this PR, I just wanted to mention it so if it's tracked we're good 🟢

File: frontend/src/components/pages/student/WriteAssessmentPage.tsx (hunk 11 above)
Comment (joyce-shi) [reply frontend/src/components/pages/student/WriteAssessmentPage.tsx]: i like

File: frontend/src/components/assessments/student-experience/SubmitButton.tsx (hunk 4 above)
Comment (joyce-shi) [reply frontend/src/components/assessments/student-experience/SubmitButton.tsx:74]: oh awesome sounds good! thank you 😊
